### DISCUSSION TITLE
Embedding and Qdrant ?

### DISCUSSION THREAD

**Discussion by jerpic** [marcominerva/ChatGptNet#148] [Ideas]
Hello,

Any plan to store and retrieve embeddings into the qdrant vector database?

Regards,

**Comment by marcominerva** [comment]
Saving and retrieving embeddings in a Vector Database is beyond the scope of this library. If you need this feature, I suggest you to take a look at the Kernel Memory library: https://github.com/microsoft/kernel-memory

**Reply by jerpic** [reply]
Thank you for your quick reply. Does it mean I can use your library with this one (kernel-memory)? My objective is to create a chat based on my own datas.

**Reply by marcominerva** [reply]
You can use Kernel Memory at all. Take also a look at the following example I made:

- https://github.com/marcominerva/OpenAIEmbeddingSample
- https://github.com/marcominerva/KernelMemoryService

**Reply by jerpic** [reply]
Hello @marcominerva,

Thank you for sharing the links, very interesting. The examples use Semantic Kernel and not ChatGptNet, you confirm that I can use Kernel Memory and ChatGptNet?

**Reply by marcominerva** [reply]
Yes, you can safely use ChatGptNet instead of Semantic Kernel.

**Reply by jerpic** [reply]
Hello @marcominerva,
Honestly, I'm struggling to find the way for using ChatGptNet instead of Semantic Kernel in your example, method "AskQuestionAsync" from the ApplicationMemoryService class. Any help would be very appreciated ;-) 

`public async Task<MemoryResponse?> AskQuestionAsync(Question question, double minimumRelevance = 0.76, string? index = null)
    {
        // Reformulate the following question taking into account the context of the chat to perform keyword search and embeddings:
        var reformulatedQuestion = await chatService.CreateQuestionAsync(question.ConversationId, question.Text);

        // Ask using the embedding search via Kernel Memory and the reformulated question.
        // If tags are provided, use them as filters with OR logic.
        var answer = await memory.AskAsync(reformulatedQuestion, index, filters: question.Tags.ToMemoryFilters(), minRelevance: minimumRelevance);

        if (answer.NoResult == false)
        {
            // If the answer has been found, add the interaction to the chat, so that it will be used for the next reformulation.
            await chatService.AddInteractionAsync(question.ConversationId, question.Text, answer.Result);

            var response = new MemoryResponse(answer.Result, answer.RelevantSources);
            return response;
        }

        return null;
    }`
Regards,

**Reply by marcominerva** [reply]
You need to modify the [ChatService.cs](https://github.com/marcominerva/KernelMemoryService/blob/master/KernelMemoryService/Services/ChatService.cs) file using `IChatGptClient` instead of `IMemoryCache` and `IChatCompletionService`.

**Reply by jerpic** [reply]
Thank you for your reply. 
Sorry to bother you. I thought that the ChatService class was there only to reformulate the question and not to chat about the documents stored into the KernelMemory. 
If I change the implementation of ChatService class (for ChatGptNet), how can I use ChatGptNet to stream the answer (coming from `var answer = await memory.AskAsync(reformulatedQuestion, index, filters: question.Tags.ToMemoryFilters(), minRelevance: minimumRelevance);
`)?

**Reply by marcominerva** [reply]
Unfortunately, at this moment Kernel Memory does not support streaming response.
